### PR TITLE
[SPARK-7173][YARN] Add label expression support for application master

### PR DESCRIPTION
**Pull request by jerryshao**
Add label expression support for AM to restrict it runs on the specific set of nodes. I tested it locally and works fine.

@sryza and @vanzin please help to review, thanks a lot.

### CODE REVIEW

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by vanzin** [comment]
LGTM.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -225,7 +225,32 @@ private[spark] class Client(
...
-    appContext.setResource(capability)
+
+    if (sparkConf.contains("spark.yarn.am.nodeLabelExpression")) {
+      try {
+        val amRequest = Records.newRecord(classOf[ResourceRequest])
+        amRequest.setResourceName(ResourceRequest.ANY)
+        amRequest.setPriority(Priority.newInstance(0))
+        amRequest.setCapability(capability)
+        amRequest.setNumContainers(1)
+        amRequest.setRelaxLocality(true)
```
Comment (sryza) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Is this required?  It should be the default.

Comment (jerryshao) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Yeah, I checked, seems not required, I will remove this line.

**Comment by zzcclp** [comment]
It's cool. Will this pr merge into 1.6?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Hi @jerryshao , could you make that change?

Comment (jerryshao) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Sorry for the delay, will change it soon.

**Comment by vanzin** [comment]
Merging to master / 1.6.

**Pull request closed without merging**